modules/raft/src/main/java/org/apache/ignite/internal/raft/RetryContext.java

```diff
@@ -69,11 +72,19 @@ class RetryContext {
      * Creates a context.
      *
      * @param targetPeer Target peer to send the request to.
+     * @param originDescription Supplier describing the origin request from which this one depends, or returning {@code null}
+     *     if this request is independent.
      * @param requestFactory Factory for creating requests to the target peer.
      * @param stopTime Timestamp that denotes the point in time up to which retry attempts will be made.
      */
-    RetryContext(Peer targetPeer, Function<Peer, ? extends NetworkMessage> requestFactory, long stopTime) {
+    RetryContext(
+            Peer targetPeer,
+            Supplier<String> originDescription,
```

Comment:
This is to avoid calculation of the description for the 'no timeout happened' case as the description is only needed for 'timeout happened' case. It seems best to avoid this calculation on the hot path as it might be noticeable.